src/java/org/apache/cassandra/io/sstable/StorageHandler.java

```diff
@@ -146,20 +147,43 @@ public StorageHandler(TableMetadataRef metadata, Directories directories, Tracke
 
     public static StorageHandler create(TableMetadataRef metadata, Directories directories, Tracker dataTracker)
     {
-        // local keyspaces are always local so don't use the remote handler even if it has been configured
-        if (remoteStorageHandler == null || Schema.isKeyspaceWithLocalStrategy(metadata.keyspace))
-            return new DefaultStorageHandler(metadata, directories, dataTracker);
+        // if legacy config is present, use it
+        if (remoteStorageHandler != null)
+            return createRemoteHandler(metadata, directories, dataTracker, remoteStorageHandler);
 
-        Class<StorageHandler> factoryClass =  FBUtilities.classForName(remoteStorageHandler, "Remote storage handler");
+        return remoteStorageHandlerFactory == null
+               ? StorageHandlerFactory.INSTANCE.create(metadata, directories, dataTracker)
+               : createRemoteHandlerWithFactory(metadata, directories, dataTracker, remoteStorageHandlerFactory);
+    }
+
+    @Deprecated
+    private static StorageHandler createRemoteHandler(TableMetadataRef metadata, Directories directories, Tracker dataTracker, String handler)
+    {
+        Class<StorageHandler> factoryClass =  FBUtilities.classForName(handler, "Remote storage handler");
 
         try
         {
             return factoryClass.getConstructor(TableMetadataRef.class, Directories.class, Tracker.class)
-                    .newInstance(metadata, directories, dataTracker);
+                               .newInstance(metadata, directories, dataTracker);
+        }
+        catch (NoSuchMethodException | IllegalAccessException | InstantiationException | InvocationTargetException e)
+        {
+            throw new ConfigurationException("Unable to find correct constructor for " + handler, e);
+        }
+    }
+
+    private static StorageHandler createRemoteHandlerWithFactory(TableMetadataRef metadata, Directories directories, Tracker dataTracker, String handlerFactory)
+    {
+        Class<StorageHandlerFactory> factoryClass =  FBUtilities.classForName(handlerFactory, "Remote storage handler factory");
+
+        try
+        {
+            StorageHandlerFactory factory = factoryClass.getConstructor().newInstance();
```

Comment:
It's a bit weird to create a create a new factory every time we want to use it, and I could see that easily be both inconvenient (means the factory cannot cache everything for instance) and unexpected (a factory that only ever create one object ...).

I'd create the factory statically, and then use that repeatedly to get a `StorageHandler`. The code here can still handle `factory == null` to means "use the legacy option".

As a minor aside, the patch seems to do "use legacy if present, otherwise check new method", and it would be more natural to me to do "use new method if present, otherwise check legacy (and maybe warn that it's deprecated)".  Admittedly, this is probably only used by cndb and we know how to handle that so this doesn't really matter but ...

Reply:
good point. I don't think the old `REMOTE_STORAGE_HANDLER` is used anywhere else. I will remove it.